### DISCUSSION TITLE
Error creating vector store from exception when executing agent (.../step10_assistant_tool_file_search.py)

### DISCUSSION THREAD

**Discussion by selfishark** [microsoft/semantic-kernel#9002] [Q&A] [answered]
Please what should be the resource to make the python/samples/getting_started_with_agents/step10_assistant_tool_file_search.py run?
- venv: ```Python 3.10; semantic-kernel==1.9.0; openai==1.43.0```
- LLM: Azure OpenAI

I get the following error:
- semantic_kernel.exceptions.agent_exceptions.AgentExecutionException: Error creating vector store
- Traceback:
```
Traceback (most recent call last):
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\semantic_kernel\agents\open_ai\open_ai_assistant_base.py", line 564, in create_vector_store
    vector_store = await self.client.beta.vector_stores.create(file_ids=file_ids)
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\resources\beta\vector_stores\vector_stores.py", line 375, in create
    return await self._post(
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\_base_client.py", line 1816, in post
    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\_base_client.py", line 1510, in request
    return await self._request(
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\_base_client.py", line 1611, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.NotFoundError: Error code: 404 - {'error': {'code': '404', 'message': 'Resource not found'}}

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "D:\CodeBase Docs\New folder\ai_agent\agent\agent_test_file_reading.py", line 76, in <module>
    asyncio.run(main())
  File "C:\Users\ltwansi\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\ltwansi\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 641, in run_until_complete
    return future.result()
  File "D:\CodeBase Docs\New folder\ai_agent\agent\agent_test_file_reading.py", line 49, in main
    agent = await AzureAssistantAgent.create(
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\semantic_kernel\agents\open_ai\azure_assistant_agent.py", line 288, in create
    vector_store_id = await agent.create_vector_store(file_ids=file_search_file_ids)
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\semantic_kernel\agents\open_ai\open_ai_assistant_base.py", line 567, in create_vector_store
    raise AgentExecutionException("Error creating vector store.") from ex
semantic_kernel.exceptions.agent_exceptions.AgentExecutionException: Error creating vector store.
```

I would appreciate any help you can provide.

Best,

**Comment by moonbox3** [comment]
Hi @selfishark, this looks to be related to an incompatible Azure OpenAI model deployment / Azure OpenAI API version. In the first stack trace, a 404 is present. In the second stack trace, it says unable to create the vector store.

I've just tested running it, and I am using a gpt-4o-mini deployment (model version 2024-07-18) with an AZURE_OPENAI_API_VERSION = "2024-08-01-preview". Can you try adjusting your model and settings, please?

Per Azure's documentation, it looks like even a 2024-05-01-preview API VERSION would work... https://learn.microsoft.com/en-us/azure/ai-services/openai/assistants-reference?tabs=python

**Reply by selfishark** [reply]
Thank you for this prompt reply @moonbox3. 
I tested my current deployment gpt-4o (model version 2024-05-13) with the following API VERSIONS but none worked
```AZURE_OPENAI_API_VERSION="2024-05-01-preview" , "2024-08-01-preview" and "2024-02-15-preview".```

I will try to get more quota to deploy gpt-4o-mini or try another resource/region and confirm my findings.

**Reply by moonbox3** [reply]
Thanks for the update, @selfishark. Will wait to hear from you after trying the other model or possibly the new region.

**Reply by selfishark** [reply]
Dear @moonbox3, 
I updated my API endpoint (and version), AI model and region to: ```region: 'east-us' - deployment: 'gpt-4o-mini' - version: '2024-08-01-preview'``` and now I face a different error: ```openai.APIStatusError: Unsupported media type```. 

(Just to confirm, I used the same resource file 'employees.pdf' and another sample '.txt' and '.json' file with no success.)

Hopefully, it's some kind of progress; here is the new traceback:
```
Traceback (most recent call last):
  File "D:\CodeBase Docs\New folder\ai_agent\agent\agent_test_file_reading.py", line 98, in <module>
    asyncio.run(main())
  File "C:\Users\ltwansi\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\ltwansi\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 641, in run_until_complete
    return future.result()
  File "D:\CodeBase Docs\New folder\ai_agent\agent\agent_test_file_reading.py", line 62, in main
    agent = await AzureAssistantAgent.create(
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\semantic_kernel\agents\open_ai\azure_assistant_agent.py", line 281, in create
    file_id = await agent.add_file(file_path=file_path, purpose="assistants")
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\semantic_kernel\agents\open_ai\open_ai_assistant_base.py", line 533, in add_file
    file = await self.client.files.create(file=file, purpose=purpose)  # type: ignore
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\resources\files.py", line 400, in create
    return await self._post(
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\_base_client.py", line 1816, in post
    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\_base_client.py", line 1510, in request
    return await self._request(
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\openai\_base_client.py", line 1611, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.APIStatusError: Unsupported media type
```
Besides, when looking for solutions I found this pending issue: https://github.com/openai/openai-python/issues/1512 - if it can help?

Thank you again for your time and consideration

**Reply by moonbox3** [reply]
This is very strange, @selfishark. I am using the same Azure region, model, and API version and I cannot repro your error. Could you try and pull the latest SK Python 1.10.0 package and see if you still experience the issue? That package version has an upgrade OpenAI dependency now at 1.47.0 instead of 1.43.0.

Are you changing anything related to the creation of the AzureAssistantAgent? Passing in new kwargs that aren't present in the sample? 

As an alternative, if you don't want to upgrade to SK 1.10.0, with your current code can you try:

```python
agent = await AzureAssistantAgent.create(
            kernel=kernel,
            service_id=service_id,
            name=AGENT_NAME,
            instructions=AGENT_INSTRUCTIONS,
            enable_file_search=True,
            vector_store_filenames=[pdf_file_path],
            deployment_name=<your deployment>,  # new
            endpoint=<your endpoint>,  # new
            api_version="2024-08-01-preview",  # new
        )
```

If you are using a .env file within VSCode and you have the .env file at the Python root project, then VSCode will load your .env vars as environment variables, and the Agent Pydantic settings will be created with the proper deployment_name, endpoint, and api_version. If you don't have it configured like this, then explicitly passing them into the agent creation could be a good troubleshooting step.

**Reply by selfishark** [reply]
@moonbox3, with direct assignment of the ```api-key, deployment_name, endpoint and api_version="2024-08-01-preview"``` 
and an upgrade of my module, ```python -m pip install --upgrade semantic-kernel[all] openai==1.47.0```;
I get another error now ```Run failed with status: `failed` for agent `FileSearch` and thread `thread_ZqpKKvVDC0a7zi5s0TzUnyCM```:

```
# user: 'Who is the youngest employee?'
Traceback (most recent call last):
  File "D:\CodeBase Docs\New folder\ai_agent\agent\agent_test_file_reading.py", line 98, in <module>
    asyncio.run(main())
  File "C:\Users\ltwansi\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\ltwansi\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 641, in run_until_complete
    return future.result()
  File "D:\CodeBase Docs\New folder\ai_agent\agent\agent_test_file_reading.py", line 88, in main
    await invoke_agent(agent, thread_id=thread_id, input="Who is the youngest employee?")
  File "D:\CodeBase Docs\New folder\ai_agent\agent\agent_test_file_reading.py", line 44, in invoke_agent
    async for content in agent.invoke(thread_id=thread_id):
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\semantic_kernel\agents\open_ai\open_ai_assistant_base.py", line 626, in invoke
    async for is_visible, content in self._invoke_internal(
  File "D:\CodeBase Docs\New folder\ai_agent\.venv\lib\site-packages\semantic_kernel\agents\open_ai\open_ai_assistant_base.py", line 725, in _invoke_internal
    raise AgentInvokeException(
semantic_kernel.exceptions.agent_exceptions.AgentInvokeException: Run failed with status: `failed` for agent `FileSearch` and thread `thread_ZqpKKvVDC0a7zi5s0TzUnyCM`
```
I don't know what is going on... I might just try to leverage this capability inside an agent configuration that already works.

**Reply by moonbox3** [reply]
Hmmm. Thanks for the additional info @selfishark. I also don't know what is going on. :( Do you get the run failed with status `failed` every 100% of the time? Can you run the assistant agent without file search? Can you run the code interpreter with a file id provided? Trying to understand at what point the error is hit.

**Reply by selfishark** [reply]
@moonbox3 sorry about the delay. - Here are my answers to the previous question:
- Do you get the run failed with status failed every 100% of the time? : Yes, unfortunately. - When I pass the env variable in plain text, the code enters the ```# user: 'Who is the youngest employee?``` as the last traceback shows otherwise nothing else just errors such as ```openai.APIStatusError: Unsupported media type```
- Can you run the assistant agent without file search? : Yes, some other instances of the agent samples work such as chatting or leveraging plugins but file search and code interpreter do not seem to work on my end.
- Can you run the code interpreter with a file id provided? : I have not tried that yet as I assumed a file needs to be created/uploaded at first?

I am attaching the traceback with logging details to this message so you can have a better insight.

[traceback.txt](https://github.com/user-attachments/files/17234181/)

**Reply by moonbox3** [reply]
@selfishark thanks for the extra information. I do want to help you get to the root cause of this. It looks like it's going to take a bit more investigation. Can we please try the following:

1. Create a new Python venv.
2. Pull the latest SK Python package (we just release 1.11.0). This will contain a little more error logging during a failed thread run.
3. In the step10_... sample, add `import logging` and then `logging.basicConfig(level=logging.DEBUG)` if you don't have it already. DEBUG is going to give us a lot more information to work with (we can see the content types being uploaded and such). 
4. Please NOTE, this is super important -- you're going to have to scrub those logs and remove your Bearer token, if it is included (it is included in mine even though I am using Entra auth). 

I think with this new debug log information, it will help me understand better where the failure is occurring.

If you want, you can email me a text file of your log directly to evan.mattson@microsoft.com.

**Reply by selfishark** [reply]
@moonbox3 I am pleased to say the code is working now:

1. I updated ```semantic-kernel to 1.11.0``` and ```openai to 1.47.0``` and used ```gpt35-turbo16k```, ```api_version="2024-05-01-preview"``` 
2. I was able to ```read a text file and pdf file``` but the ```AI failed to read the content of other types of files``` such as Python code.

I think my other problem is how my ```.env``` is populated in the code; For some reason when the key, endpoint, etc are passed in plain text it works, so I am going to follow up on that with a new environment as suggested.

I hope I can still email you with my findings or other queries.

**Reply by moonbox3** [reply]
Hi @selfishark, glad to hear you're making some progress! For #2, are you trying to have the model read python code from a .py file that was uploaded to the vector store? Try including the code file for the code interpreter to use.

Yes, please do follow up via email with detailed (scrubbed) logs if you continue to hit issues.

**Reply by moonbox3** [reply]
One other note -- you shouldn't need to install a specific OpenAI 1.47.0 dependency. Semantic Kernel has this package dependency and will install it for you. The code may experience issues (potentially what you've been hitting?) if you have different package versions installed / env related issues.

**Reply by selfishark** [reply]
@moonbox3 thank you I'll carry on and keep you informed. 

Would you please clarify this bit?

- "Try including the code file for the code interpreter to use."

**Reply by moonbox3** [reply, marked as answer]
@selfishark there are two ways to have the OpenAI Assistant handle files -- via the vector store (like you've been working with recently) or via the code interpreter (as shown in this example: https://github.com/microsoft/semantic-kernel/blob/ae29d8735ad2c756135fa2a9de4baa70b250c760/python/samples/concepts/agents/assistant_agent_file_manipulation.py#L62C9-L63C52.py#L63). By passing in a file to the code interpreter it is able to use it during its analysis.

**Reply by selfishark** [reply]
@moonbox3 I sent you an email. - Thank you again for your support! I am really greatfull for your time and consideration.
